src/librustc/middle/expr_use_visitor.rs

```diff
@@ -558,24 +558,27 @@ impl<'a, 'gcx, 'tcx> ExprUseVisitor<'a, 'gcx, 'tcx> {
             }
             ty::TyError => { }
             _ => {
-                let def_id = self.mc.tables.type_dependent_defs()[call.hir_id].def_id();
-                let call_scope = region::Scope::Node(call.hir_id.local_id);
-                match OverloadedCallType::from_method_id(self.tcx(), def_id) {
-                    FnMutOverloadedCall => {
-                        let call_scope_r = self.tcx().mk_region(ty::ReScope(call_scope));
-                        self.borrow_expr(callee,
-                                         call_scope_r,
-                                         ty::MutBorrow,
-                                         ClosureInvocation);
-                    }
-                    FnOverloadedCall => {
-                        let call_scope_r = self.tcx().mk_region(ty::ReScope(call_scope));
-                        self.borrow_expr(callee,
-                                         call_scope_r,
-                                         ty::ImmBorrow,
-                                         ClosureInvocation);
+                let type_dependent_defs = self.mc.tables.type_dependent_defs();
+                if type_dependent_defs.contains_key(call.hir_id) {
```

Comment:
Could you use this instead? It's more standard.
```Rust
if let Some(def_id) = type_dependent_defs.get(call.hir_id)
```

Reply:
I would actually prefer

```Rust
let def_id = match type_dependent_defs.get(call.hir_id) {
    Some(def) => def.def_id(),
    None => tcx.sess.delay_span_bug(call.span, "no type-dependant def for overloaded call?)
};
```

as that would make sure we get an ICE if there's ever a mismatch.

Reply:
Isn't an ICE exactly what we're trying to avoid in this situation? In the issue this fix addresses, this function is called for a struct (with no type-dependent call definition) — an error has already been logged, we just want to make sure we don't attempt anything else with it.